Proposal: Per-extension language preferences (#641)
hober/web-extensions: proposals/per-extension-language-preferences.md

# Proposal: Per-extension language preferences

**Summary**

This proposal allows developers (users) to set a specific language for their extension (which may be different from the browser or system default language) and create a language selection menu for users.

**Document Metadata**

**Author:** hanguokai

> **Sponsoring Browser:** Chromium

Reviewer: I forget if we had already agreed this - was there a discussion somewhere? If not I will take it to the team.

Author: @oliverdunk [a meeting note](https://github.com/w3c/webextensions/blob/658a794ac59f014e77a932593950c4188e2ca3a4/_minutes/2024-05-09-wecg.md?plain=1#L135) said this.

Reviewer: Cool, let me confirm then. I definitely like the change, I'm just being extra careful any time we add ourselves as the sponsoring browser since we agreed on some strict rules about that meaning we will implement in a timely fashion.

Reviewer: Hey Jackie - I spoke to the team about this. We're supportive of the idea, but it is very unlikely we would be able to implement this ourselves any time soon. With that in mind, we'd only be comfortable sponsoring this if there was an external contributor able to implement it.

@xeenon / @Rob--W With that in mind, would either Apple or Mozilla be more likely to implement this soon and be interested in sponsoring?

[Android 13](https://developer.android.com/guide/topics/resources/app-languages), [iOS 13 and macOS Catalina](https://developer.apple.com/news/?id=u2cfuj88) support a feature
called "Per-app language preferences (settings)". This proposal brings the same function to browser extensions.

#### Use Cases

- Extensions create a language selection menu for users.
- [Optional] Browsers could potentially provide a built-in language selection menu for users, just like Android and iOS.
- [Optional] Extensions could integrate third-party i18n implementations (frameworks) with the browser's built-in language selection menu.

### Known Consumers

This is a generic i18n feature, which means that all types of extensions can use it. Some extensions already provide this functionality via workarounds or non-browser.i18n implementations.

## Specification

### Definitions

In this document:
- **locale** or **language code** or **language tag** is a string that represent a language, defined in [BCP 47](https://www.rfc-editor.org/info/bcp47).

> For example, `en-US`, `zh-CN`, `fr`. It is used by `browser.i18n`, `Date`, `Intl` and various other APIs.

Reviewer 2: `zh-CN` is presumably meant to indicate Simplified Chinese, which is also used in Singapore. It might be better to use `zh-Hans` as the language tag in the example.

See also https://www.w3.org/International/articles/language-tags/index.en.html#script

Author: Perhaps due to historical reasons, browsers still uses 'zh-CN' instead of 'zh-Hans'. For example, `navigator.language` returns "zh-CN" in Chrome and Firefox on macOS. And Chrome only supports [a list of languages](https://developer.chrome.com/docs/extensions/reference/api/i18n#locales) in the `/_locales/` directory, which only contains a limited combination of languages and regions.

I think solving this problem seems to go beyond the core issue that this proposal aims to address. This proposal is intended to expand functionality on the basis of existing capabilities.

Reviewer 3: Still would be good and is also relevant for this issue what locales a browser supports / can deal with. See also: https://github.com/w3c/webextensions/issues/131

For example, the Naver Whale store rejects any language tag which includes the script subtag (Hans, Latn, Cyrl, and others).

- **the user preferred language** is the language set by `i18n.setCurrentLanguage(code)`. The user may not have set it.
- **the extension displayed language** is the language which the extension is displayed in. Note that the extension displayed language may be different from the browser UI language. For example, if the user preferred language is not set, the browser UI language is English, and the extension only supports French (default) and Japanese, then the extension displayed language is French because `i18n.getMessage()` returns French message at this situation.

### Schema

```ts
/**
 * Get the extension displayed language.

>  * return the language tag, e.g. en-US.

Reviewer 3: ```suggestion
 * return a [BCP47 language tag](https://www.rfc-editor.org/bcp/bcp47.html), e.g. en-US.
```

Author: In the "Definitions" section, I already defined it. So don't need add it everywhere else.

Reviewer 3: @hanguokai Considering the different formatting used, as mentioned here, https://github.com/w3c/webextensions/issues/642, I believe it is good to be explicit here. Your call though.

 */
i18n.getCurrentLanguage() : string

/**
 * Set the user preferred language.
 * if code is null, revert to the unset state.

>  * if code is not valid, return a rejected Promise.

Reviewer 3: ```suggestion
 * if code is not valid, for example when it uses an underscore, return a rejected Promise.
```

Documenting https://github.com/w3c/webextensions/pull/641#discussion_r1647699428

 * else return a Promise resolved when the operation is complete.
 */

> i18n.setCurrentLanguage(code: string) : Promise<void>

Reviewer 3: Should this method accept codes with underscores? See also https://github.com/w3c/webextensions/issues/642 To align with the web Intl standards I suggest we throw an exception in this case.

Author: No.

/**
 * Get all languages that supported by this extension.
 * return a Promise resolved with an array of language tags.
 */

> i18n.getAllLanguages(): Promise<string[]>

Reviewer: I'm not sure if this one makes sense, given that extensions should already know the languages they support. I think this gets tricky in certain situations (partial strings, parent language tags) so I wonder if we should leave this out. I see that having it provides some convenience but I'm not sure it's essential for the MVP.

Author: > extensions should already know the languages they support

Yes. It is just a convenient method. And I think it is easy to implement.

> in certain situations (partial strings, parent language tags)

_Languages with regional subtags_ are not a problem for users. For example, a language selection menu that includes French, French (Canada), French (Belgium) and French (France). This method should return all languages if they are there.

For partial strings, from the platform's standpoint, the platform does not need to consider these, but assumes that all these languages are supported.

Reviewer 3: @oliverdunk That is to say the browser per definition supports all the language tags specified in the extension. I suggest we change the method name to `i18n.getAvailableLanguages` to let the browser return all the languages the extension can call `setCurrentLanguage` to.

It is not just out of convenience. For example, for the Whale Store I have to exclude language tags with script variations. (zh-Hant, sr-Latn). Thus those are removed from `_locales`. Having to specify a list of all supported languages for each browser and store in every area of the extension this is used is not as straight forward as it might sound.

Reviewer 4: It's important to know the difference between a _locale_ and _localization_ here. There can be many locales supported by a single localization. This API seems to want to return the list of available localizations, but users typically need to choose between the list of available locales. The locale can be used do to internationalized formatting/processing (such as in MessageFormat or using `Intl`), which provides a better-adapted, richer localized experience than merely translating the static messages.

For example, many applications come in just two varieties of English (US vs. UK/International English), but support many locales. The language negotiation/resource fallback (such as used by `getMessage`) takes care of filling in the localized strings for any requested locale (including using the default language when the requested locale has no available localization), but you still want the locale (in most cases) to provide processing/formatting. Note that many applications also tailor the fallback, so that, for example, the `es-419` (Spanish, Latin America) localization serves many seemingly-unrelated (from a tag perspective) Spanish locales (`es-CO`, `es-CR`, `es-AR`, etc.....)

Author: > There can be many locales supported by a single localization. This API seems to want to return the list of available localizations

Yes, this proposal and the current `browser.i18n.getMessage()` mechanism ( [MDN doc](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Internationalization) , [Chrome doc](https://developer.chrome.com/docs/extensions/reference/api/i18n) ) mainly focus on localization (language translations). In other words, it focuses on languages, not all possible regions.

Regarding expanding locales, I think this may be an area that needs to be improved in the future, or expand by developers themselves.

Reviewer 4: > Regarding expanding locales, I think this may be an area that needs to be improved in the future, or expand by developers themselves.

I agree about alternate fallback paths (although `Intl` has a proposal for this and it is already the way that CLDR data--and thus `Intl`--works for some locale-based "best match" selection vs. BCP47 Lookup), but think that this document isn't putting enough thought into the separation of locales and localizations/translations. Trying to use the same mechanism for both will disappoint users. Trivial examples with only a couple of languages don't match the needs of those developers who need to support a fairly large set of locales.

`getMessage` tries to use a BCP47 Lookup type of fallback to select the correct message "like a resource bundle" (cf. Java or GNU gettext for examples of other resource bundle systems). Resource selection often had a dual fallback capability (the _messages_ files and the specific keys within each message file).

The resource files want to use the **_least_** specific locale as their identifier possible (the better to provide coverage for many locales). The user, however, wants to specify the **_most_** specific locale possible (including script, region, and various locale extensions where applicable), the better to tailor the runtime experience.

In this document, there are calls to things such as "selecting which _locale_" to use (i.e. a picker) as well as "selecting which _localization_" to use (i.e. which language file or files to download. These lists want to be different (the list of locales is long, while the list of localizations is typically shorter).

---

Note that there is on-going work at W3C related to manifest (they are adopting slightly different mechanisms for managing localization). Also, I am the editor of [Developing Localizable Manifests](https://www.w3.org/TR/localizable-manifests/), which tries to enumerate a lot of this material and which readers on this thread might find useful. Lastly, there is work at Unicode on [MessageFormat 2](https://github.com/unicode-org/message-format-wg/) and proposed for MessageResource (which is similar to the message files in `getMessage`)

Reviewer: @aphillips, thanks for taking your time to share your expertize on this issue. I am certainly out of my depth so the extra context and pointers on what to consider are really appreciated. If you'd like to attend a public meeting at some point, we could likely carve out some time to make sure we can have an in depth discussion. Otherwise your time continuing to help with the conversation here is valued regardless.

Based on all of the above, I am still of the opinion that offering a `getAllLanguages` that is intended for rendering a menu is going to be very hard to get right. Given users want to choose a fairly specific locale, it would likely need to return a much longer list of anything you can `setCurrentLanguage` to, as @carlosjeurissen suggested. However, that seems like it would be a long list which would be hard to render with the right nesting / structure without additional work. You would also need a mapping from returned codes to strings which are human readable.

I can see the value in a function that returns "which of the folders in _locales were accepted". Perhaps we could lean into that with an API like `getParsedLocalesDirectories()`? Based on @aphillips' explanation it seems like locales may be the wrong word there but I think we are slightly backed into a corner by the existing usage of `_locales` as the canonical place for extensions to store messages.json files.

For `setCurrentLanguage`, I'm convinced after this discussion that you should be able to set it to any valid locale, however specific, and the browser should pick the appropriate language file with fallbacks if needed. That way you can offer a fairly detailed picker for users and rely on the browser to use its built-in fallback behavior.

Author: > rendering a menu is going to be very hard to get right …… 

Creating a language selection menu has always been a challenge in i18n work, but it is also a necessary work. It can be very simple or very complex. For example, the extension can only list a few languages without variations (like regions and scripts), or list languages with limited or all regions, or let users set languages, regions and extensions (like calendar, date format, etc) separately.

> You would also need a mapping from returned codes to strings which are human readable.

`Intl.DisplayNames` can help it. It also depends on how developers want to design it. For example, a language code can be displayed in two display names (e.g. "zh-CN" can be displayed by "Chinese Simplified (简体中文)"). Anyway, display names are not what this proposal is trying to solve.

Author: > For setCurrentLanguage, I'm convinced after this discussion that you should be able to set it to any valid locale, however specific, and the browser should pick the appropriate language file with fallbacks if needed. That way you can offer a fairly detailed picker for users and rely on the browser to use its built-in fallback behavior.

@oliverdunk Do you know why Chrome only allow developers to use [a fixed list of languages](https://developer.chrome.com/docs/extensions/reference/api/i18n#locales)? It is limited by the browser or Chrome Web Store? If it is a hard limitation, that means the browser doesn't support some locales to use in `setCurrentLanguage()` and the `/_locales/` directory.

Reviewer: @hanguokai My understanding is that it is based on the locales supported by some of the internal libraries we use to handle i18n.

Reviewer 3: As far as I know both the chrome web store and chrome browser support more languages than specified in that fixed list of languages. The Chrome Web Store also supports and recognises more languages currently than this list contains.

@oliverdunk are there any current internals which would restrict the languages one could use in `setCurrentLanguage()` to this fixed list?

For `i18n.getAllLanguages()` specifically I have opened a separate PR to specify and discuss this method as it is useful outside the `i18n.setCurrentLanguage()` use cases and could be implemented separately. See https://github.com/w3c/webextensions/pull/942

> i18n.getAllLanguages(): Promise<string[]>

Reviewer 3: Awaiting resolution of https://github.com/w3c/webextensions/issues/642, we might want to consider returning an object with more than just the code. Think about something like:
```json
[{
  "code": "en-US",
  "path": "en_US",
  "current": true,
  "default": true
}]
```

Author: I think it is unnecessary, it complicates the design.

/**
 * After changing to a new language, the browser triggers a language changed event.
 * The callback is function(code: string) : void.
 * The code parameter in the callback is the new language.
 */
i18n.onLanguageChanged.addListener(callback)
```

### Behavior

#### Behavior of `i18n.getCurrentLanguage()`

This method return the language that the extension is displayed in.

- If the extension doesn't use `browser.i18n` (there is no "_locales" directory), return `undefined`.

> - If the preferred language is not set by `i18n.setCurrentLanguage()`, returns the current language used by `i18n.getMessage()`, assuming that all languages support all possible keys.

Reviewer 3: ```suggestion
- If the preferred language has not been set by `i18n.setCurrentLanguage()`, returns the first language for which a message file exists following the same fallback mechanism used by `i18n.getMessage()
```

As a sidenote, the fallback mechanism is not the same across browsers. See: https://github.com/w3c/webextensions/issues/296

- If the preferred language is set by `i18n.setCurrentLanguage()`, and the extension supports this language, then return this language.

> - If the preferred language is set by `i18n.setCurrentLanguage()`, but the extension doesn't support this language (no message file for this language), then treat as if the preferred language is not set. This is an edge case, for example, the language was removed when the extension was upgraded.

Reviewer: Should we allow you to get into this state? I would've expected `setCurrentLanguage` to throw an error and abort updating the language.

Author: It depends on the implementation.  Although `setCurrentLanguage()` throw an error, but as I said in the "Implementation Notes" section: When an extension is upgraded, the browser should check to see if the languages it supports has changed (especially in the case of deletion). If the browser does this check, it will not get into this state, otherwise it may happen.

From a specification perspective, I'm just listing this possibility. From an implementation perspective, browsers just need to make sure that this problem can be avoided.

Reviewer: As mentioned above, following the discussion with @aphillips I'm more in favour of allowing this now. Perhaps we should still throw in extreme cases (if you try to set the language to French, but you only have English message files) but beyond that allowing you to use browser fallbacks seems helpful.

> - If the preferred language is set by `i18n.setCurrentLanguage()`, but the extension doesn't support this language (no message file for this language), then treat as if the preferred language is not set. This is an edge case, for example, the language was removed when the extension was upgraded.

Reviewer 3: ```suggestion
- If the preferred language has been set by `i18n.setCurrentLanguage()`, and the extension supports this language, then return this language.
- If the preferred language has been set by `i18n.setCurrentLanguage()`, but the extension doesn't support this language (no message file for this language), then treat as if the preferred language is not set. This is an edge case, for example, the language was removed when the extension was upgraded.
```

Changing "is set" to "has been set" to prevent the confusion mentioned here: https://github.com/w3c/webextensions/pull/641/files#r1644443944

Author: Thanks. Grammatical issues aren't the top priority, though.

Reviewer 3: @hanguokai The motivation is not improving the grammar but to improve the meaning of the sentences. In this case to prevent confusion found in https://github.com/w3c/webextensions/pull/641#discussion_r1644443944. Generally I very much agree with leaving the grammar and formatting to later in the process. For example, with the icon_variants proposal it was counter productive to early-on add line-breaks for formatting motivations as with every change this had to be updated.

##### Use Case 1: use the extension displayed language with other locale-related APIs.

For example, if the preferred language is not set, the language of the browser UI is English (en-US), the extension supports French (default) and Japanese, then `i18n.getCurrentLanguage()` return `"fr"` because `i18n.getMessage()` return French message at this situation.

```js
// use current extension displayed language to format date
const locale = browser.i18n.getCurrentLanguage();
const today_date = new Date().toLocaleString(locale); // format date string in 'fr'

// if don't use current extension displayed language, it may be formatted in other locale
const today_date = new Date().toLocaleString(); // format date string in 'en-US'
```

##### Use Case 2: show the current language in the extension language selection menu.

#### Behavior of `i18n.setCurrentLanguage(code)`

This method sets the user preferred language for this extension.

> - If the extension doesn't use `browser.i18n` (there is no "_locales" directory), return a rejected Promise.

Reviewer: Do we need this line? I think the line below covers it and I think that aligns more closely with how this would be implemented in practice.

Author: From an implementation perspective, this is a bit redundant. But, from a specification perspective, this situation is slightly different from the following situation.

Reviewer: I think I'd still be in favour of removing this. It doesn't change anything about the behaviour and having the spec guide implementors seems like a desirable thing.

Author: I removed this line.

- If the code is an invalid language tag or an unsupported language by this extension, return a rejected Promise.
- If the code is null, revert to the unset state.
- else do the following:
  1. save the language persistently, and make sure that the language is prioritized for use next time.
  1. return a resolved Promise.
  1. trigger `i18n.onLanguageChanged` event.

In addition, when `i18n.setCurrentLanguage(code)` success:

> - the browser should update related browser UI and extension UI, because some values in manifest.json may be changed, like `name`, `short_name`, `description` and `action.default_title`.

Reviewer: I want to confirm this with the team. I suspect we might not want to translate everything, e.g certain parts of chrome://extensions, since those should show in the browser language.

Author: For browser UI (like chrome://extensions), it makes sense to use the browser UI language if the extension supports.
For extension UI, e.g. `action.default_title`, I think the browser should update it. I will update this sentence.

Author: I rephrased text about updating browser UI and extension UI in a new commit.

### Browser built-in language selection menu

Based on the capabilities provided by this proposal, browsers could provide a unified built-in language selection menu for extensions, like Android and iOS.
Whether to provide the built-in language selection menu and how to implement it is up to the browser to decide. This is just a suggestion.

The built-in language selection menu has the following benefits:
- Easy for developers: Developers do not need to implement it themselves, they only need to adapt to this proposal.
- Easy for users: A unified UI makes it easier for users to use, otherwise each extension might be set up differently.
- It is convenient for developers to develop and test i18n functions by switching the extension language.

### Integrate other i18n implementations (frameworks) with browser.i18n

This proposal can also be integrated with third-party i18n frameworks, as these frameworks typically allow developers to specify the language to be used.

> For example, developers use `i18n.getCurrentLanguage()` and `i18n.setCurrentLanguage()` for the user preferred language, but don't use `i18n.getMessage()`.

Reviewer 3: Does this mean a developer would not have to specify `messages.json` files to inform the browser what languages it supports? If so this would be inconsistent with the `i18n.getCurrentLanguage()` behaviour specified above. Would we want an alternative way to specify the supported languages by the extension in this case? A manifest key like `supported_locales` which specifies a list of BCP47 language tags stating all the languages it supports.

Author: No, they still need to add messages.json files for languages they support. Because the extension name, short name and description can only be localized with `browser.i18n` mechanism.

Reviewer 3: @hanguokai @oliverdunk How about instead of specifying `builtin_languages_menu`, we add a `available_locales` property in the manifest listing all the locales supported by the extension. This would prevent the need for specifying the messages.json as sometimes (from my experience with some extensions) a language is supported to some extend without having messages present in messages.json. To overcome this one would need to copy the english messages.json but this seems very counter-intuitive. Specifying an empty messages.json is not sufficient to at least tell the Chrome Web Store a specific language is supported.

```js
otherI18nFramework.setLanguage(browser.i18n.getCurrentLanguage());

let text = otherI18nFramework.getMessage(key);
```

### New Permissions

N/A

### Manifest File Changes

> If the browser would like to support the built-in language selection menu for extensions, the manifest file should add a new key for developers to opt-in this feature, like the following:

Reviewer: I think we should chat about this with other browser vendors, and only keep this section if at least one vendor plans to implement it.

Reviewer 3: @oliverdunk If no browser vendor wants to implement something like this, we can offer these APIs in `browser.management` so an extension can create this UI to customise the languages of other extensions.

Reviewer: What would a `browser.management` API look like? If it involves rendering new UI, I don't think Chrome would be supportive. I would expect other browsers to feel similarly if they don't plan on implementing a menu.

Author: I moved the "Browser built-in language selection menu" to the "Future Work" section in this document. When browser vendors are interested in this, we can discuss it again.

Reviewer 3: @oliverdunk the whole point of adding APIs to `browser.management` would be to prevent the browser having to render UI but let extensions figure this out. imagine something like:
```js
browser.management.setLanguage(extensionId, locale)
```

To let an extension change the language of another extension.

Reviewer: Ah, I missed this. I could go either way on if that should be in scope here but that seems reasonable.

```
{
  "name": "__MSG_extName__",
  "default_locale": "en",

>   "builtin_languages_menu": true | false(default)

Reviewer 3: Why would we want this to be an opt-in feature? The only downside I see here is changing the language using a built-in languages menu could lead to extension pages in a dual language setup.

Author: Because some extensions are not ready to adapt to it. For example, some extensions seem to support many languages, but these are just for adding listing descriptions in other languages in the extension store (Extensions Store only allows adding descriptions in languages that the extension supports),  they don't actually support these languages in the extension.

### Schema

```ts
/**
 * Get the extension displayed language.
 * return the language tag, e.g. en-US.
 */
i18n.getCurrentLanguage() : string

/**
 * Set the user preferred language.
 * if code is null, revert to the unset state.
 * if code is not valid, return a rejected Promise.
 * else return a Promise resolved when the operation is complete.
 */
i18n.setCurrentLanguage(code: string) : Promise<void>

/**
 * Get all languages that supported by this extension.
 * return a Promise resolved with an array of language tags.
 */
i18n.getAllLanguages(): Promise<string[]>

/**
 * After changing to a new language, the browser triggers a language changed event.
 * The callback is function(code: string) : void.
 * The code parameter in the callback is the new language.
 */
i18n.onLanguageChanged.addListener(callback)
```

> ### Behavior

Reviewer 5: The `i18n` API is one of the few APIs that are available to all extension contexts, including lower-privileged content scripts.

Could you specify the intended availability of the API?

I'd say, the default is to NOT expose the APIs to content scripts unless a good reason is given otherwise.

The `getCurrentLanguage` method sounds potentially useful to content scripts, but I'm not sure what the value of this would be. It is specified to be synchronous, which means that browser implementations are forced to make the language choice available to all content processes. It is likely for this info to already be available along with the messages though.

`setCurrentLanguage` - To prevent (compromised) content processes from affecting extensions, I would restrict its availability to extension contexts only (e.g. background script, popup, extension tabs, ...).

`getAllLanguages` - given that its primary intended use is to determine the language to change to, I'd also restrict this to non-content scripts too.

`i18n.onLanguageChanged` - sounds potentially useful to content scripts. But even without this, extensions could use extension messaging to notify content scripts upon detecting a language change.

Author: For content scripts, there are several possibilities:

1. Loosest: all methods are available in content scripts.
1. Strictest: all methods are not available.
1. Some are not available.

I'm leaning towards option 3. Specifically, `setCurrentLanguage`, `getCurrentLanguage` and `getAllLanguages` are not available in content scripts (at least `setCurrentLanguage` is not available). And `onLanguageChanged` is available.

Reviewer 3: @hanguokai I would suggest `getCurrentLanguage` to be available in content scripts considering the extension might want to use the current language code outside getMessage situations. In addition, if we add some predefined i18n message, something like @@current_locale, this would be available using getMessage in contentScripts anyway.

Reviewer: If `onLanguageChanged` is available, I definitely agree with @carlosjeurissen that the corresponding getter should be available too.

Author: I add a new section of Content Scripts Availability

- If the preferred language is not set by `i18n.setCurrentLanguage()`, it is consistent with the existing behavior.
- If the preferred language is set by `i18n.setCurrentLanguage()`, and the extension supports this language, then prioritize using that language.
- If the preferred language is set by `i18n.setCurrentLanguage()`, but the extension doesn't support this language, then treat as if the preferred language is not set. This is an edge case, for example, the language was removed when the extension was upgraded.

### Integrate other i18n implementations (frameworks) with browser.i18n

This proposal can also be integrated with third-party i18n frameworks, as these frameworks typically allow developers to specify the language to be used.
For example, developers use `i18n.getCurrentLanguage()` and `i18n.setCurrentLanguage()` for the user preferred language, but don't use `i18n.getMessage()`.

```js
otherI18nFramework.setLanguage(browser.i18n.getCurrentLanguage());

let text = otherI18nFramework.getMessage(key);
```

### New Permissions

N/A

### Manifest File Changes

N/A

## Security and Privacy

### Exposed Sensitive Data

N/A

### Abuse Mitigations

N/A

> ### Additional Security Considerations

Reviewer 5: This proposal has `N/A` at "Abuse Mitigations" and "Additional Security Considerations".

I see potential for abuse: An extension could rapidly change their name at runtime to make it more difficult for users to discover what extension they are interacting with. The proposal does not describe any mitigations for this abuse concern.

Reviewer 3: @Rob--W In general I believe it would be best to fetch the translations for the manifest file in the language of the containing UI. So if it will be used in the browser extension management page, it would match the language of said page. Same for the context menu items on the action button. This would prevent the issue you are describing.

Author: Although I don't think this poses a real security problem. However, the browser can limit the frequency of `setCurrentLanguage` method calls, which is up to the browser.

Author: I added  a supplementary note in the Abuse Mitigations section. let me know if it's Ok.

### Schema

```ts
/**
 * Get the extension displayed language.
 * return the language tag, e.g. en-US.
 */
i18n.getCurrentLanguage() : string

/**
 * Set the user preferred language.
 * if code is null, revert to the unset state.
 * if code is not valid, return a rejected Promise.
 * else return a Promise resolved when the operation is complete.
 */
i18n.setCurrentLanguage(code: string) : Promise<void>

/**
 * Get all languages that supported by this extension.
 * return a Promise resolved with an array of language tags.
 */
i18n.getAllLanguages(): Promise<string[]>

/**
 * After changing to a new language, the browser triggers a language changed event.
 * The callback is function(code: string) : void.
 * The code parameter in the callback is the new language.
 */
i18n.onLanguageChanged.addListener(callback)
```

### Behavior

#### Content Scripts Availability

> In history, all `browser.i18n` APIs were available in content scripts. But some of methods in this proposal should not be available in content scripts because of their purpose.

Reviewer 5: ```suggestion
Unlike other APIs, all `browser.i18n` APIs are historically available to content scripts. To counter abuse, some of the features in this proposal should not be available in content scripts.
```

> | Method  | Content Scripts Availability |

Reviewer 5: ```suggestion
| API  | Content Scripts Availability |
```

Method -> API because onLanguageChanged is not a method.

| ------------- | ------------- |
| `getCurrentLanguage()` | Yes |
| `setCurrentLanguage()` | No |
| `getAllLanguages()` | No |

> | `onLanguageChanged()` | Yes |

Reviewer 5: ```suggestion
| `onLanguageChanged` | Yes |
```

This is an event, not a method.

#### A New Predefined Message `@@current_locale`
There is an exsiting predefined message `@@ui_locale`, that reflects the value of `i18n.getUILanguage()`, but the value uses underscore (e.g. "en_US") as separator in Chrome and hyphen (e.g. "en-US") in Firefox.

> Relative to `@@ui_locale`, a new predefined message, ``@@current_locale``, should be added, which reflects the value of `i18n.getCurrentLanguage()`.

Reviewer 5: This is a late addition to the proposal. What use cases are served with the proposed `@current_locale`? This property does not automatically follow from the three listed use cases in the proposal document.

Additionally, because the stylesheet does not update automatically in response to UI language changes, the use of the proposed `@current_locale` in the stylesheet can result in UI getting out of sync with the actual language, unless the extension has additional code to deal with language changes (with `i18n.onLanguageChanged`). If an extension already needs to account for this, then they may as well do all language-dependent CSS from JavaScript. These days, they can use CSS variables to control stylesheet behavior in a more powerful way than `@`-macros can. For example, the only way for an extension to reload the stylesheet to read the new `@` values is to replace the style sheet element with a new one that has the same URL and a random extra addition (query string) to bust the style cache.

Because of this, I am inclined to vote for removing the proposed `@current_locale` from the proposal.

Author: Personally, I don't pursue this predefined message very much, but only add it for completeness reason. @carlosjeurissen Would you mind deleting this part? Because you suggested it.

Reviewer 3: Also not feeling too strongly about it. One of the motivations is described here:
https://github.com/w3c/webextensions/issues/642#issuecomment-2181397837

As a way to get the path of the language files in JavaScript for example.

In general I agree with not using the predefined messages in CSS. Personally not been using them in any shape or form in the past few years in any projects in CSS. It very much feels like a hack and we might want to deprecate some of them moving forward.

- If the preferred language is not set by `i18n.setCurrentLanguage()`, it is consistent with the existing behavior.
- If the preferred language is set by `i18n.setCurrentLanguage()`, and the extension supports this language, then prioritize using that language.
- If the preferred language is set by `i18n.setCurrentLanguage()`, but the extension doesn't support this language, then treat as if the preferred language is not set. This is an edge case, for example, the language was removed when the extension was upgraded.

### Integrate other i18n implementations (frameworks) with browser.i18n

This proposal can also be integrated with third-party i18n frameworks, as these frameworks typically allow developers to specify the language to be used.
For example, developers use `i18n.getCurrentLanguage()` and `i18n.setCurrentLanguage()` for the user preferred language, but don't use `i18n.getMessage()`.

```js
otherI18nFramework.setLanguage(browser.i18n.getCurrentLanguage());

let text = otherI18nFramework.getMessage(key);
```

### New Permissions

N/A

### Manifest File Changes

N/A

## Security and Privacy

### Exposed Sensitive Data

N/A

### Abuse Mitigations

> A case of abuse is changing the extension name/description dynamically by `i18n.setCurrentLanguage()`. This is not a security threat because all metadata published in the extension store is audited. And for the extension management page (e.g. chrome://extensions/), the extension name should be displayed in the same language as the management page (i.e. the Browser UI language), so it is not affected by `i18n.setCurrentLanguage()`.

Reviewer 5: > "This is not a security threat because all metadata published in the extension store is audited."

Please drop this part. This claim about it not being a security threat is not true. A legitimate extension can have different names for branding purposes, and that would be accepted by the web store. This is indistinguishable from a malicious extension that intentionally chooses different names to maximize confusion, when the extension switches names dynamically.

Here is a rephrased suggestion that calls out the exact risk, and also language on the mitigation. I also added another example of a browser UI surface where this matters (permissions dialog/prompt).

```suggestion
A case of abuse is changing the extension name/description dynamically by `i18n.setCurrentLanguage()` in an attempt to confuse the user and hinder their ability to identify the extension. A mitigation to this issue is to ignore changes from `setCurrentLanguage` in the extension management parts of the browser UI, and always use the browser's configured UI language instead. This includes extension management pages (e.g. `chrome://extensions` in Chrome and `about:addons` in Firefox) and permission dialogs.
```
